### DISCUSSION TITLE
How do I order by RAND?

### DISCUSSION THREAD

**Discussion by alper** [SeaQL/sea-orm#1009] [Q&A] [answered]
I have a query like this:

```
let venue = Venues::find()
    .filter(entity::venues::Column::GoogleName.eq(""))
    .order_by(entity::venues::Column::Id, sea_orm::Order::Asc)
    .one(sqlite_conn)
    .await
    .unwrap();
```

How would I turn this into something that orders by random?

**Comment by billy1624** [comment, marked as answer]
Hey @alper, hi there! This is more of a SQL related question. But yeah, you can order the result in random order with `ORDER RAND()` / `ORDER RANDOM()` depending on which db you're on.

- https://www.mysqltutorial.org/select-random-records-database-table.aspx
- https://www.educba.com/postgresql-order-by-random/
- https://stackoverflow.com/questions/1253561/sqlite-order-by-rand

We don't have `RANDOM` function in SeaQuery yet. So, you'd have to go with `Expr::cust()`.

However, a small PR would be appreciated to introduce the support of `RANDOM` function. Add a `RANDOM` variant to https://github.com/SeaQL/sea-query/blob/cb6fd12ecec848cfe90f05ad7f2a27f2db25028c/src/func.rs#L8-L27

**Reply by alper** [reply]
OK. Good to know.

> This is more of a SQL related question.

I could lookup very easily what the SQL should look like here but it wasn't immediately obvious to me how to translate it into Sea.

I'll take a look at this and also see if I can add the `RAND` there.

**Comment by studentenherz** [comment]
Hi @alper, I see you added the random function [here](https://github.com/SeaQL/sea-query/pull/430), thanks. I still don't know how would I use that in `.order_by`. Would you mind explaining?

**Reply by billy1624** [reply]
E.g.

```rs
entity::find().order_by_expr(Func::random(), Order::Asc).all(db).await?
```

https://docs.rs/sea-orm/latest/sea_orm/query/trait.OrderedStatement.html#method.order_by_expr

**Reply by rbozan** [reply]
@billy1624 It's not clear to me.

```
error[E0599]: no method named `order_by_expr` found for struct `sea_orm::Select` in the current scope
  --> src/node_generator.rs:79:10
   |
76 | /     database_entities::tip::Entity::find()
77 | |         .filter(database_entities::tip::Column::CourseId.eq(course_id))
78 | |         .limit(2)
79 | |         .order_by_expr(Func::random(), Order::Asc)
   | |         -^^^^^^^^^^^^^ method not found in `Select<Entity>`
   | |_________|
   | 
```


```rs
    database_entities::tip::Entity::find()
        .filter(database_entities::tip::Column::CourseId.eq(course_id))
        .limit(2)
        .order_by_expr(Func::random(), Order::Asc)
        .all(&db)
        .await?;
```

Only these methods are allowed:
![image](https://github.com/SeaQL/sea-orm/assets/7997154/19eedba1-69a5-4ffb-9120-67224da05908)

**Reply by YiNNx** [reply]
The code below works for me, using sea_orm::query:

```rust
let select = Entity::find().as_query().to_owned().order_by_expr(SimpleExpr::FunctionCall(Func::random()), Order::Asc) .to_owned();
let statement = self.db().get_database_backend().build(&select);
CustomResultStruct::find_by_statement(statement).one(self.db()).await?
```

**Reply by joebnb** [reply]
> E.g.
> 
> ```rust
> entity::find().order_by_expr(Func::random(), Order::Asc).all(db).await?
> ```
> 
> https://docs.rs/sea-orm/latest/sea_orm/query/trait.OrderedStatement.html#method.order_by_expr

is there have some hidden tricks, im not find `order_by_expr ` when after entity::find()

-----updated

i fond a smiple way to work around

```rust
  query_builder = Entity::find().order_by(
            sea_orm::sea_query::SimpleExpr::FunctionCall(Func::random()),
            sea_orm::Order::Desc,
        );
```

because order_by permit  pass a IntoSimpleExpr implemented struct
```rust
pub trait QueryOrder: Sized {
    type QueryStatement: OrderedStatement;

    // Required method
    fn query(&mut self) -> &mut SelectStatement;

    // Provided methods
    fn order_by<C>(self, col: C, ord: Order) -> Self
       where C: IntoSimpleExpr { ... }
    fn order_by_asc<C>(self, col: C) -> Self
       where C: IntoSimpleExpr { ... }
    fn order_by_desc<C>(self, col: C) -> Self
       where C: IntoSimpleExpr { ... }
    fn order_by_with_nulls<C>(
        self,
        col: C,
        ord: Order,
        nulls: NullOrdering,
    ) -> Self
       where C: IntoSimpleExpr { ... }
}
```

[a comment by an automated account is not shown]
